Index: epan/dissectors/packet-bt-dht.c
===================================================================
--- epan/dissectors/packet-bt-dht.c	(revision 50714)
+++ epan/dissectors/packet-bt-dht.c	(working copy)
@@ -142,16 +142,15 @@
 {
   guint start_offset;
 
-  start_offset = offset;
-
   /* we have confirmed that the first byte is 'i' */
   offset += 1;
+  start_offset = offset;
 
   while( tvb_get_guint8(tvb,offset)!='e' )
     offset += 1;
 
-  *result = tvb_get_ephemeral_string( tvb, offset, offset-start_offset-1 );
-  proto_tree_add_string_format( tree, hf_bencoded_int, tvb, offset, offset-start_offset-1, *result,
+  *result = tvb_get_ephemeral_string( tvb, start_offset, offset-start_offset);
+  proto_tree_add_string_format( tree, hf_bencoded_int, tvb, start_offset, offset-start_offset, *result,
     "%s: %s", label, *result );
 
   offset += 1;
